### PR TITLE
8325064: C2 SuperWord: refactor construct_bb

### PR DESCRIPTION
**Pull request by eme64**
Subtask of https://github.com/openjdk/jdk/pull/16620

The goal is to further disentangle different "components" in Superword.

In this refactoring, I disentangle the `bb`, `reduction` and `memory_slice` "components" which were all intertwined in `construct_bb`.

1. Move memory slice code -> `analyze_memory_slices`.
2. Remove reduction checking code -> simply use the `is_marked_reduction_loop` condition outside.
3. `_data_entry`: was used for non-CFG nodes in the loop that have no input node that is also inside the loop. But that actually never happens! I removed that array, and replaced the code with verification.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8325064](https://bugs.openjdk.org/browse/JDK-8325064): C2 SuperWord: refactor construct_bb (**Sub-task** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [aaaaae63](https://git.openjdk.org/jdk/pull/17657/files/aaaaae6379991290c6b1424f698e9551fe385cf0)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [aaaaae63](https://git.openjdk.org/jdk/pull/17657/files/aaaaae6379991290c6b1424f698e9551fe385cf0)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17657/head:pull/17657` \
`$ git checkout pull/17657`

Update a local copy of the PR: \
`$ git checkout pull/17657` \
`$ git pull https://git.openjdk.org/jdk.git pull/17657/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17657`

View PR using the GUI difftool: \
`$ git pr show -t 17657`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17657.diff">https://git.openjdk.org/jdk/pull/17657.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17657#issuecomment-1919549201)

### CODE REVIEW

File: src/hotspot/share/opto/superword.cpp [hunk 1]
```diff
@@ -896,6 +918,36 @@ void SuperWord::dependence_graph() {
...
+#ifndef PRODUCT
+void SuperWord::print_memory_slices() {
+  tty->print_cr("\nSuperWord::print_memory_slices: %s",
+                _mem_slice_head.length() > 0 ? "" : "NONE");
+  for (int m = 0; m < _mem_slice_head.length(); m++) {
+    tty->print("%6d ", m);  _mem_slice_head.at(m)->dump();
+    tty->print("       ");  _mem_slice_tail.at(m)->dump();
+  }
+}
+#endif
```
Comment (eme64) [inline src/hotspot/share/opto/superword.cpp:949]: Note: both of these methods are refactored out of `construct_bb`

File: src/hotspot/share/opto/superword.cpp [hunk 2]
```diff
@@ -2950,177 +3002,110 @@ bool SuperWord::is_vector_use(Node* use, int u_idx) {
...
-        if (!found) {
-          assert(n != entry, "can't be entry");
-          _data_entry.push(n);
-        }
-      }
-    }
-  }
-
-  // Find memory slices (head and tail)
-  for (DUIterator_Fast imax, i = lp()->fast_outs(imax); i < imax; i++) {
```
Comment (eme64) [inline src/hotspot/share/opto/superword.cpp:2994]: Note: moved the memory slice code to `analyze_memory_slices`

File: src/hotspot/share/opto/superword.cpp [hunk 3]
```diff
@@ -2950,177 +3002,110 @@ bool SuperWord::is_vector_use(Node* use, int u_idx) {
...
-          if (def && in_bb(def)) {
+          if (def != nullptr && in_bb(def)) {
             found = true;
             break;
           }
         }
-        if (!found) {
-          assert(n != entry, "can't be entry");
-          _data_entry.push(n);
-        }
```
Comment (eme64) [inline src/hotspot/share/opto/superword.cpp:2988]: Note: `found` is always true, it turns out. I added an assert. And this means I can also remove `_data_entry`, as we now never push anything to it.

File: src/hotspot/share/opto/superword.cpp [hunk 4]
```diff
@@ -2950,177 +3002,110 @@ bool SuperWord::is_vector_use(Node* use, int u_idx) {
...
     } else if (!post_visited.test(bb_idx(n))) {
       // cross or back arc
+      int old_length = stack.length();
       for (DUIterator_Fast imax, i = n->fast_outs(imax); i < imax; i++) {
-        Node *use = n->fast_out(i);
+        Node* use = n->fast_out(i);
         if (in_bb(use) && !visited.test(bb_idx(use)) &&
             // Don't go around backedge
-            (!use->is_Phi() || n == entry)) {
-          if (is_marked_reduction(use)) {
```
Comment (eme64) [inline src/hotspot/share/opto/superword.cpp:3040]: Note: I would like to remove this reduction code from the basic block code. It separates the different "components".

Instead, we can just check `is_marked_reduction_loop` outside, which checks if we have any marked reduction.
The only difference is that we don't do the implemented check. But most platforms implement the reductions, and we check that again later in `SuperWord::implemented` anyway.

File: src/hotspot/share/opto/superword.cpp [hunk 5]
```diff
@@ -2950,177 +3002,110 @@ bool SuperWord::is_vector_use(Node* use, int u_idx) {
...
-  if (is_trace_superword_memory_slices()) {
-    tty->print_cr("\nmemory slices: %s", _mem_slice_head.length() > 0 ? "" : "NONE");
-    for (int m = 0; m < _mem_slice_head.length(); m++) {
-      tty->print("%3d ", m); _mem_slice_head.at(m)->dump();
-      tty->print("    ");    _mem_slice_tail.at(m)->dump();
-    }
   }
 #endif
-  assert(rpo_idx == -1 && bb_ct == _block.length(), "all block members found");
-  return (_mem_slice_head.length() > 0) || (reduction_uses > 0) || (_data_entry.length() > 0);
```
Comment (eme64) [inline src/hotspot/share/opto/superword.cpp:3094]: Note: Moved this condition out:
```
  if (!is_marked_reduction_loop() &&
      _mem_slice_head.is_empty()) {
```

File: src/hotspot/share/opto/superword.cpp [hunk 6]
```diff
@@ -2950,177 +3002,110 @@ bool SuperWord::is_vector_use(Node* use, int u_idx) {
...
-  }
-  // Set value
-  _block.at_put_grow(n_pos, n);
-  _node_info.at_put_grow(n_pos, SWNodeInfo::initial);
-  // Adjust map from node->_idx to _block index
-  for (int i = n_pos; i < _block.length(); i++) {
-    set_bb_idx(_block.at(i), i);
-  }
-}
-
```
Comment (eme64) [inline src/hotspot/share/opto/superword.cpp:3123]: Note: not used anymore, already prevous to this change

**Review by vnkozlov — APPROVED** [review]
This looks reasonable.

**Review by chhagedorn — APPROVED** [review]
Looks good!

**Pull request closed without merging**